Provide support for using multiple MAX31865 sensors

brvdboss opened the pull request:
## Description:

**Related issue (if applicable):** fixes #<Tasmota issue number goes here>

## Checklist:
  - [x] The pull request is done against the latest dev branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR.
  - [x] The code change is tested and works on Tasmota core ESP8266 V.2.7.3.2
  - [x] The code change is tested and works on core ESP32 V.1.12.2
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

Note; for the ESP32-test, I did do al the builds and those compile. Did not test on actual hardware yet.
Did test on actual hardware for ESP8266

This is the implementation of what was discussed in https://github.com/arendst/Tasmota/issues/9086

brvdboss committed:
Provide support for using multiple MAX31865 sensors

arendst commented on tasmota/language/bg_BG.h:
```diff
@@ -622,6 +622,11 @@
+#define D_SENSOR_MAX31865_CS2  "MX31865 CS2"
+#define D_SENSOR_MAX31865_CS3  "MX31865 CS3"
+#define D_SENSOR_MAX31865_CS4  "MX31865 CS4"
+#define D_SENSOR_MAX31865_CS5  "MX31865 CS5"
```
Don't do this. Look at how buttons or relays are implemented. Waste of translation space.

arendst commented on tasmota/tasmota_template.h:
```diff
@@ -337,7 +342,8 @@ const char kSensorNames[] PROGMEM =
   D_SENSOR_IEM3000_TX "|" D_SENSOR_IEM3000_RX "|"
-  D_SENSOR_ZIGBEE_RST
+  D_SENSOR_ZIGBEE_RST "|"
+  D_SENSOR_MAX31865_CS1 "|" D_SENSOR_MAX31865_CS2 "|" D_SENSOR_MAX31865_CS3 "|" D_SENSOR_MAX31865_CS4 "|" D_SENSOR_MAX31865_CS5
```
See first comment

arendst commented on tasmota/tasmota_template_ESP32.h:
```diff
@@ -238,7 +243,8 @@ const char kSensorNames[] PROGMEM =
   D_SENSOR_IEM3000_TX "|" D_SENSOR_IEM3000_RX "|"
-  D_SENSOR_ZIGBEE_RST
+  D_SENSOR_ZIGBEE_RST "|"
+  D_SENSOR_MAX31865_CS1 "|" D_SENSOR_MAX31865_CS2 "|" D_SENSOR_MAX31865_CS3 "|" D_SENSOR_MAX31865_CS4 "|" D_SENSOR_MAX31865_CS5
```
See previous comment

arendst commented on tasmota/tasmota_template_ESP32.h:
```diff
@@ -528,6 +534,13 @@ const uint16_t kGpioNiceList[] PROGMEM = {
+  GPIO_SSPI_MAX31865_CS3, //MAX31865 CS selectable pin 3
+  GPIO_SSPI_MAX31865_CS4, //MAX31865 CS selectable pin 4
+  GPIO_SSPI_MAX31865_CS5, //MAX31865 CS selectable pin 5
+#endif
```
Needs AGPIO macro. See jusr before your entries

arendst commented on tasmota/xsns_47_max31865.ino:
```diff
@@ -35,71 +35,137 @@
 }
 
 /*
 *   MAX31865_GetResult(void)
```
too much redundancy.  Good to know: tasmota is all about less code doing more

arendst commented on tasmota/xsns_47_max31865.ino:
```diff
@@ -35,71 +35,137 @@
-          DomoticzSensor(DZ_TEMP, temperature);
-        }
+          if (0 == tele_period) {
+            DomoticzSensor(DZ_TEMP, temperature);
```
Only one temperature can be handled by domoticz and knx. Look at other multisensor implementatioms

arendst commented on tasmota/xsns_47_max31865.ino:
```diff
@@ -35,71 +35,137 @@
-          temperature, resistance, MAX31865_Result.ErrorCode);
+        if(Json){
+          ResponseAppend_P(PSTR(",\"MAX31865(%d)\":{\"" D_JSON_TEMPERATURE "\":%s,\"" D_JSON_RESISTANCE "\":%s,\"" D_JSON_ERROR "\":%d}"), \
+            i, temperature, resistance, MAX31865_Result[i].ErrorCode);
```
Do not use brackets a sindex. Lok how other multisensor drivers are doing it like bh1750 allowing either a dash or underline as index.

arendst commented:
As the total amount of gpios is nearing 250.as being the max on legacy esp8266 I'm reluctant to add 5 new ones while the usecase is specific you. I'm more thinking of reusing gpios defined as switch. I'll need to dive in if this is even possible.

brvdboss commented:
Thanks for all the feedback!  I'll improve the pull request based on your feedback and by looking at the other drivers you referenced. I will likely only be able to do this next weekend

brvdboss committed:
resolve first batch of feedback:

- fix language files to use more efficient translation
- adjust template files to use more efficient translation
- include AGPIO macro for ESP32
- use IndexSeparator

arendst committed:
Add MAX38165 GPIO's the ESP32 (and hopefully soon the ESP8266) way

arendst commented:
Stay open.

arendst commented:
Thx. Merged with changes.

kobusdbr commented:
Hello all. I am working with a brewery here in south Africa. They use PT100 temperature sensors in their fermenters. I would like to add a D1 Mini to each fermenter so that they can read and display temperature data. Want to use HA.  We will then use those sensors to switch relays to switch on solenoids to cool the brew. The Home assistant thermostat will work for this. I have found some info on using PT100 with arduino but nothing for tasmota as yet. Has anyone got some experience with this that are able to point me in the right direction?

lalo-uy commented:
It is suported, but not included in default bin.
You will need to compile yourself with

define USE_MAX31865


El El lun, 17 de may. de 2021 a la(s) 06:25, kobusdbr <
***@***.***> escribió:

> Hello all. I am working with a brewery here in south Africa. They use
> PT100 temperature sensors in their fermenters. I would like to add a D1
> Mini to each fermenter so that they can read and display temperature data.
> Want to use HA. We will then use those sensors to switch relays to switch
> on solenoids to cool the brew. The Home assistant thermostat will work for
> this. I have found some info on using PT100 with arduino but nothing for
> tasmota as yet. Has anyone got some experience with this that are able to
> point me in the right direction?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/pull/9103#issuecomment-842169659>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACXBW4LC5VB5WKPEU7QP7MDTODOJNANCNFSM4QAKZO4A>
> .
>

bryanb334 commented:
I see the max31855 listed in the sensor bin.  If I want to use muntiple max31855/thermocouples so I still need to complile my own binary?  Thanks

barbudor commented:
You are confusing MAX31855 and MAX31865
The first one is indeed included in the sensor bin but support only 1 device
The 2nd one (the one this PR is about) is not included by default and support up to 6 devices.

bryanb334 commented:
Thanks for your reply.

On Sun, Apr 16, 2023, 12:03 AM Barbudor, ***@***.***> wrote:

> You are confusing MAX31855 and MAX31865
> The first one is indeed included in the sensor bin but support only 1
> device
> The 2nd one (the one this PR is about) is not included by default and
> support up to 6 devices.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/pull/9103#issuecomment-1510130411>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANHN5YGK5ZSSS4MEAU73FXLXBOKVNANCNFSM4QAKZO4A>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
